test/common/tls/cert_validator/default_validator_test.cc

```diff
@@ -429,32 +429,38 @@ TEST(DefaultCertValidatorTest, TestMatchSubjectAltNameIncorrectTypeMatched) {
   EXPECT_FALSE(DefaultCertValidator::matchSubjectAltName(cert.get(), subject_alt_name_matchers));
 }
 
-TEST(DefaultCertValidatorTest, TestMatchSubjectAltNameWildcardDNSMatched) {
+TEST(DefaultCertValidatorTest, TestMatchSubjectAltNameExactDNSFailure) {
+  // This will only be tested under debug (assertion is triggered) to ensure
+  // that the class will not be initialized in the DNS-exact mode.
+#ifndef NDEBUG
   NiceMock<Server::Configuration::MockServerFactoryContext> context;
 
+  envoy::type::matcher::v3::StringMatcher matcher;
+  matcher.set_exact("api.example.com");
+  EXPECT_DEATH(std::make_unique<StringSanMatcher>(GEN_DNS, matcher, context),
```

Comment:
Idea that isn't required for this PR: create an `EXPECT_ASSERT` macro to avoid `#ifndef` sprinkled throughout tests

Reply:
I've been thinking about this, but I think there are 2 minor challenges:
1. perf: If the test dies at the end, why execute it at all in non-debug mode.
2. if the test needs to ensure it can run in debug and non-debug mode the statements following the expect_death will only work in one of the modes.

It may be possible to mark these tests with a SKIP_TEST_NDEBUG(); macro that will skip running a test that has an assertion failure.
LMK what you think, or if there's a better alternative, and I can try to get a PR out later.

Reply:
It would probably only work if the entire test body was in the macro, and the entire thing is skipped when assertions are not compiled in.